pyro/ops/gaussian.py

```diff
@@ -34,28 +35,176 @@ def __init__(self, log_normalizer, info_vec, precision):
+
+    def event_permute(self, perm):
+        """
+        Permute along event dimension.
+        """
+        assert isinstance(perm, torch.Tensor)
+        assert perm.shape == (self.dim(),)
+        info_vec = self.info_vec[..., perm]
+        precision = self.precision[..., perm][..., perm, :]
+        return Gaussian(self.log_normalizer, info_vec, precision)
+
+    def __add__(self, other):
+        """
+        Adds two Gaussians in log-density space.
+        """
+        assert isinstance(other, Gaussian)
+        assert self.dim() == other.dim()
+        return Gaussian(self.log_normalizer + other.log_normalizer,
+                        self.info_vec + other.info_vec,
+                        self.precision + other.precision)
+
     def log_density(self, value):
         """
-        Evaluate the log density of this Gaussian at a point value.
+        Evaluate the log density of this Gaussian at a point value::
 
-            `-0.5 * value.T @ precision @ value + value.T @ info_vec + log_normalizer`
+            -0.5 * value.T @ precision @ value + value.T @ info_vec + log_normalizer
 
         This is mainly used for testing.
         """
         if value.size(-1) == 0:
             batch_shape = broadcast_shape(value.shape[:-1], self.batch_shape)
             return self.log_normalizer.expand(batch_shape)
-        result = (-0.5) * torch.matmul(self.precision, value.unsqueeze(-1)).squeeze(-1)
+        result = (-0.5) * self.precision.matmul(value.unsqueeze(-1)).squeeze(-1)
         result = result + self.info_vec
         result = (value * result).sum(-1)
         return result + self.log_normalizer
 
+    def condition(self, value):
+        """
+        Condition this Gaussian on a trailing subset of its state.
+        This should satisfy::
+
+            g.condition(y).dim() == g.dim() - y.size(-1)
+
+        Note that since this is a non-normalized Gaussian, we include the
+        density of ``y`` in the result. Thus :meth:`condition` is similar to a
+        ``functools.partial`` binding of arguments::
+
+            left = x[..., :n]
+            right = x[..., n:]
+            g.log_density(x) == g.condition(right).log_density(left)
+        """
+        assert isinstance(value, torch.Tensor)
+        assert value.size(-1) <= self.info_vec.size(-1)
+
+        n = self.dim() - value.size(-1)
+        info_a = self.info_vec[..., :n]
+        info_b = self.info_vec[..., n:]
+        P_aa = self.precision[..., :n, :n]
+        P_ab = self.precision[..., :n, n:]
+        P_bb = self.precision[..., n:, n:]
+        b = value
+
+        info_vec = info_a - P_ab.matmul(b.unsqueeze(-1)).squeeze(-1)
+        precision = P_aa
+        log_normalizer = (self.log_normalizer +
+                          -0.5 * P_bb.matmul(b.unsqueeze(-1)).squeeze(-1).mul(b).sum(-1) +
+                          b.mul(info_b).sum(-1))
+        return Gaussian(log_normalizer, info_vec, precision)
+
+    def marginalize(self, slice_):
```

Comment:
@fehiepsi I believe this is the final op we need to normalize `GaussianGaussianMRF`. I'm working on the linear algebra now. I've added the two equations below as unit tests. Any help appreciated 🙂